pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/ParseLock.java

```diff
@@ -27,49 +29,65 @@ public void ensureParsed() {
             LOG.trace("{} {}: {}", Thread.currentThread().getName(), String.format("%-15s", prefix), this);
         }
     }
 
+    void checkWeAreNotParsingAnother() {
+       // by default do nothing, overridden by CheckedParseLock
+    }
+
+
+    void releaseLock() {
+        // by default do nothing, overridden by CheckedParseLock
+    }
+
     private ParseStatus getFinalStatus() {
         ParseStatus status = this.status;
         if (!status.isFinished) {
             logParseLockTrace("waiting on");
 
-            synchronized (this) {
-                logParseLockTrace("locked");
-
-                status = this.status;
-                if (status == ParseStatus.NOT_PARSED) {
-                    this.status = ParseStatus.BEING_PARSED;
-                    try {
-                        boolean success = doParse();
-                        status = success ? ParseStatus.FULL : ParseStatus.FAILED;
-                        finishParse(!success);
-                    } catch (Throwable t) {
-                        status = ParseStatus.FAILED;
-                        LOG.error("Parsing failed in ParseLock#doParse() of {}", name, t);
-                        finishParse(true);
+            try {
+                checkWeAreNotParsingAnother();
+                synchronized (this) {
+                    logParseLockTrace("locked");
+
+
+                    status = this.status;
+                    if (status == ParseStatus.NOT_PARSED) {
+                        this.status = ParseStatus.BEING_PARSED;
+                        try {
+                            boolean success = doParse();
+                            status = success ? ParseStatus.FULL : ParseStatus.FAILED;
+                            finishParse(!success);
+                        } catch (Throwable t) {
+                            status = ParseStatus.FAILED;
+                            LOG.error("Parsing failed in ParseLock#doParse() of {}", name, t);
+                            finishParse(true);
+                        }
+
+                        // the status must be updated as last statement, so that
+                        // other threads see the status FULL or FAILED only after finishParse()
+                        // returns. Otherwise, some fields might not have been initialized.
+                        //
+                        // Note: the current thread might reenter the parsing logic through
+                        // finishParse() -> ... -> ensureParsed(). In that case the status is still BEING_PARSED,
+                        // and we don't call finishParse() again. See below for canReenter()
+                        this.status = status;
+
+                        assert status.isFinished : "Inconsistent status " + status;
+                        assert postCondition() : "Post condition not satisfied after parsing sig " + this;
+                    } else if (status == ParseStatus.BEING_PARSED && !canReenter()) {
```

Comment:
The only place, where we still use canReenter is now in GenericSigBase... did you try, whether we can get rid of canReenter there too? That would make it a bit simpler, I hope...

Reply:
I tried this and the one time this behavior is used is related to type parameters. When we compute annotations on type parameters we go and ask for the typevar's upper bound, which is in the same parse lock. I will take a look to see if we can also get rid of this

Reply:
Thanks!